src/logging.c

```diff
@@ -104,47 +172,153 @@ void rcutils_logging_console_output_handler(
+    char token[RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN];
+    // Look for a token end delimiter.
+    size_t chars_to_end_delim = rcutils_find(str + i, token_end_delimiter);
+    remaining_chars = size - i;
+    if (chars_to_end_delim > remaining_chars) {
+      // No end delimiters found in the remainder of the format string;
+      // there won't be any more tokens so shortcut the rest of the checking.
+      RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+        remaining_chars, output_buffer_size, allocator, output_buffer, static_output_buffer)
+      memcpy(output_buffer + old_output_buffer_len, str + i, remaining_chars + 1);
+      break;
+    }
+    // Found what looks like a token; determine if it's recognized.
+    size_t token_len = chars_to_end_delim - 1;  // Not including delimiters.
+    memcpy(token, str + i + 1, token_len);  // Skip the start delimiter.
+    token[token_len] = '\0';
+    const char * token_expansion = NULL;
+    // The resulting token_expansion string must always be null-terminated.
+    if (strcmp("severity", token) == 0) {
+      token_expansion = severity_string;
+    } else if (strcmp("name", token) == 0) {
+      token_expansion = name;
+    } else if (strcmp("message", token) == 0) {
+      token_expansion = message_buffer;
+    } else if (strcmp("function_name", token) == 0) {
+      token_expansion = location ? location->function_name : "\"\"";
+    } else if (strcmp("file_name", token) == 0) {
+      token_expansion = location ? location->file_name : "\"\"";
+    } else if (strcmp("line_number", token) == 0) {
+      if (location) {
+        // Allow 9 digits for the expansion (otherwise, truncate). Even in the case of truncation
+        // the result will still be null-terminated.
+        char line_number_expansion[10];
+        written = rcutils_snprintf(
+          line_number_expansion, sizeof(line_number_expansion), "%zu", location->line_number);
+        if (written < 0) {
+          fprintf(
+            stderr,
+            "failed to format line number: '%zu'\n",
+            location->line_number);
+          goto cleanup;
+        }
+        token_expansion = line_number_expansion;
+      } else {
+        token_expansion = "0";
+      }
+    } else {
+      // This wasn't a token; print the start delimiter and continue the search as usual
+      // (the substring might contain more start delimiters).
+      RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+        1, output_buffer_size, allocator, output_buffer, static_output_buffer)
+      memcpy(output_buffer + old_output_buffer_len, str + i, 1);
+      output_buffer[old_output_buffer_len + 1] = '\0';
+      i++;
+      continue;
+    }
+    size_t n = strlen(token_expansion);
+    RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+      n, output_buffer_size, allocator, output_buffer, static_output_buffer)
+    memcpy(output_buffer + old_output_buffer_len, token_expansion, n + 1);
```

Comment:
Looking at this, I wonder if you might be better off using `repl_str()`: https://github.com/ros2/rcutils/blob/6d6d5507362415fd377098c0b5973b91072dfafa/include/rcutils/repl_str.h#L127-L132

Did you consider it? Was there some reason to not use it for replacing tokens like `{name}` with an actual value? (there very well might have been)

Reply:
TBH I hadn't seen it, but I've taken a look and don't think we'd want to pull it in "as is". We'd need at least 6 calls to it, and each allocates memory, while we had been trying to avoid allocation if possible in the logging calls themselves. Certainly it'd be possible to have a version of `rcutils_repl_str` that only allocates when necessary, as with `RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER`, and if/when we have another instance of needing a `rcutils_repl_str` that only allocates when necessary, I think that's what we should consider switching to

Reply:
Ok, that's fine (I thought that might be the case).

Btw, is that an opportunity for optimization, could you avoid expanding tokens which do not change over time, like the name or severity? Perhaps that's too much of a premature optimization.

In the other extreme, is it that big of a deal to allocate so many times during a logging call? I mean obviously we'd love have an allocation free logging function, but that might not be realistic unless we just truncate them to a fixed length. By comparison using a std::string in normal cases might result in many separate memory allocations and that seems to work ok in most cases despite the overhead.

By the way, the "performance" section of the website where I got the `repl_str()` function is pretty interesting:

http://creativeandcritical.net/str-replace-c#performance

I think the one we have isn't great for small strings with lots of matches, but honestly it probably doesn't matter that much. When someone needs to make giga-logs per second, they can come and help us optimize it :smile:.

Reply:
Yeah the logging is only guaranteed to be allocation free at the moment if the expanded output is <1024, it doesn't truncate (but did at some point in this PR). From the existing code I was under the impression that we valued only allocating when necessary, but your point is fair about `std::string`; I am probably taking it to the extreme and `repl_str` wouldn't be that bad. If I were to do it again from scratch I think it would be worth the extra `realloc`s for readability, but for now we can benefit from the slight optimisation. Thanks for pointing it out so I don't unnecessarily optimise for memory allocation in the future.

And I like the idea about saving the expanded token string.. we don't have the infrastructure for even caching the severity threshold just yet to avoid getting into this function unnecessarily, but once the logger concept is there for that then we might be able to avoid recalculating it